src/averageAges.js

```diff
@@ -1,61 +1,63 @@
 'use strict';
 
 /**
- * Implement calculateMenAverageAge function
- *
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
  *
  * @param {object[]} people
  * @param {number} century - optional
  *
  * @return {number}
  */
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = people.filter(person => person.sex === 'm');
+  const menInCentury = men.filter(man => Math.ceil(man.died / 100) === century);
+  const menCalculated = century ? menInCentury : men;
+  const menAge = menCalculated.map(man => man.died - man.born);
+
+  return menAge.reduce((sum, n) => sum + n) / menAge.length;
 }
 
 /**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
  *
  * @param {object[]} people
  * @param {boolean} withChildren - optional
  *
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = people.filter(person => person.sex === 'f');
+  const mothers = women.filter(woman => people
+    .find(child => child.mother === woman.name));
+  const womenCalculated = withChildren ? mothers : women;
+  const womenAge = womenCalculated.map(woman => woman.died - woman.born);
+
+  return womenAge.reduce((sum, n) => sum + n) / womenAge.length;
 }
 
 /**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
  *
  * @param {object[]} people
  * @param {boolean} onlyWithSon - optional
  *
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const children = people.filter(person => person.mother);
+  const sons = people.filter(person => person.mother && person.sex === 'm');
+  const childrenCalculated = onlyWithSon ? sons : children;
+
+  const ageDifference = [];
+
+  childrenCalculated.map(person => {
+    const mother = people.find(mom => mom.name === person.mother);
+
+    if (mother) {
+      const diff = person.born - mother.born;
+
+      ageDifference.push(diff);
+    };
+  });
```

Comment:
I think you could declare an "ageDifference" as a result of **map** method, and in the map method return **diff** variable, so it would create an array and push data into array.

Reply:
agree, fix @DenysKleimenov comment